### ISSUE TITLE
Excessive vector tile requests with modified tileUrlFunction

### ISSUE THREAD

**Issue by Biosphakus** [openlayers/openlayers#12136]
**Describe the bug**
I am using a modified `tileUrlFunction` to request vector tiles from variable paths depending on the current zoom level and context (e.g. UI selection). However, it seems that vector tiles are requested multiple times in this setup. This can be seen by returning the values calculated within the `tileUrlFunction` to the console. It is also noticeable visually: On zoom, tiles from several paths are loaded until they are overwritten by the correct tile at last. I suspect this might be one reason for the poor performance I'm experiencing when using vector tiles (cf. [my question](https://gis.stackexchange.com/questions/383735/openlayers-performance-on-satellite-hybrid-map-raster-tiles-vector-tiles-vec) at GIS Stackexchange).

Using the same `tileUrlFunction` with raster tiles works as expected, requesting each tile only once.

**To Reproduce**
1. Use a modified `tileUrlFunction` to request vector tiles depending on current zoom level and context. At each call of the function, output the parameters of the requested tile to the console. 

```
var selectedMode = ...;
function getVectorTilePath() { ... }

function loadTiles(tileCoordinate) {
    const z = tileCoordinate[0];
    const x = tileCoordinate[1];
    const y = tileCoordinate[2];

    var path = getVectorTilePath(z, selectedMode);

    console.log("z:", z, "x:", x, "y:", y, "path:", path);

    return "mytileserver/"
              + path
              + "/" + z + "/" + x + "/" + y
              + ".pbf";
    }
	
var vt = new VectorTileLayer({
              source: new VectorTileSource({
                format: new MVT(),
                tileGrid: ...,
                tileUrlFunction: loadTiles,
                zDirection: 0,
                transition: 0,
          }),
          preload: 0,
          useInterimTilesOnError: false,
          style: ...,
})	
	

var map = new Map({
    target: 'map',
    loadTilesWhileAnimating: true,
    loadTilesWhileInteracting: true,
    projection: ...,
    layers: [ vt ],
    view: ...,
});
```

2. The output of the console (after a single zoom step) indicates that tiles are requested multiple times. Tiles might be excessively requested with a different path attribute (see above). Also, completely identical tiles seem to be requested multiple times.
```
z: 3 x: 11 y: 2 path: path1 
z: 3 x: 10 y: 2 path: path1 
z: 3 x: 11 y: 2 path: path1  
z: 3 x: 10 y: 2 path: path1 
z: 2 x: 5 y: 0 path: path0 
z: 1 x: 2 y: 0 path: path0
z: 2 x: 5 y: 0 path: path0
z: 3 x: 11 y: 2 path: path1 
z: 1 x: 2 y: 0 z: 1 path: path0
z: 3 x: 11 y: 2 z: 3 path: path1
```

This is, in principle, what the output looks like, `path0` representing tiles usually requested at lower zoom levels `z` than those with `path1`. (The console output is much larger but I have modified it so that duplicates are better visible). When I use the same function with raster tiles, it works as intended without duplicate requests.

**Expected behavior**
On an interaction with the map, request each vector tile only once and using the correct path only.

**Comment by ahocevar** [comment]
I think what you're observing is not related to the custom `tileUrlFunction`. I'm not sure about the reason for the repeated requests for the same tile, but the high amount of requests for lower resolution tiles might be reduced by #12137.

**Comment by Biosphakus** [comment]
Thank you for your super quick reply! I don't know about the technical details of your pull request but I just wanted to add some specifics for my use case which I'd love to be considered: I want to avoid loading lower resolution whenever possible because they would lead to a wrong visual impression during loading because of the `tileUrlFunction`. 

By using different paths in `tileUrlFunction` I load vector tiles with slightly differently semantics. One could think of it as state borders that are loaded in different granularity (country versus district borders) depending on the user's preferences and zoom level. If lower resolution tiles are requested, a different path might be selected resulting in wrong granularity during loading. This was also the reason I was using `preload: 0` but I don't know if it has the desired effect here.

**Comment by ahocevar** [comment]
OpenLayers does not actively request lower resolution tiles, unless you set `preload` to a value greater than zero. What you're seeing is probably tile requests during zoom-in animations. And these are reduced by the pull request I referenced.

But there might be another issue regarding cache keys. To find out if that is the case, can you try with a custom tile function that always uses the same path?

**Comment by Biosphakus** [comment]
The issue seems to persist even when using a custom tile function with a fixed path. 
``` var path = "path0"; ```

For **vector tiles**, the console output still looks messy, loading much more tiles than expected (reduced example):
```
x: 8 y: 13 z: 7 path: path0
x: 8 y: 14 z: 7 path: path0
x: 9 y: 13 z: 7 path: path0
x: 9 y: 14 z: 7 path: path0
x: 8 y: 13 z: 6 path: path0
x: 8 y: 14 z: 6 path: path0
x: 9 y: 13 z: 6 path: path0 
x: 9 y: 14 z: 6 path: path0 
x: 4 y: 6 z: 5 path: path0
```

For **raster tiles**, the output looks just as expected (reduced example):
```
x: 9 y: 31 z: 7 path: path0
x: 9 y: 32 z: 7 path: path0
x: 9 y: 33 z: 7 path: path0
x: 10 y: 31 z: 7 path: path0
x: 10 y: 32 z: 7 path: path0
x: 10 y: 33 z: 7 path: path0
x: 11 y: 31 z: 7 path: path0
```
(zooming to slightly different parts of the map)

**Comment by ahocevar** [comment]
Thanks for your continued effort on this. Looks like there is a bug indeed. I have a rough idea where it could come from - the `getSourceTiles()` function in `ol/VectorRenderTile`.

I won't be able to fix this before May, so if someone else wants to take a look, a pull request would be much appreciated.

**Comment by Biosphakus** [comment]
Sorry, I'm asking again, but just to be sure if it wise to close this issue: From what I see, the second bug relating to this seems not to be solved by the merge, or is it?

**Comment by ahocevar** [comment]
Thanks, I just reopened the issue @Biosphakus. In oder to investigate this further, I need your complete source and tilegrid setup, preferrably in a code sandbox (https://codesandbox.io/).

**Comment by Biosphakus** [comment]
I have produced a minimal working example and copied the code to the codesandbox [here](https://codesandbox.io/s/kind-thompson-po1th?file=/src/index.html). For some reason, I can't get it to work within the sandbox but that might just be me being unexperienced with the intricacies of the sandbox. Feel free to change the sandbox accordingly. Also, as I'm using my own tiles (which are not avaible in the sandbox), an error is expected when tiles are to be loaded. However, what you should be able to see is the output to the console, illustrating the excess tile requests.

**Comment by mike-000** [comment]
Just because the tile URL is being calculated doesn't mean the tile is going to be loaded, and even if it is required it will most likely be cached.  Logging calls to the tileLoadFunction will show much less activity than the tileUrlFunction.

```
var defaultTileLoadFunction = vectorLayer_tiles.getSource().getTileLoadFunction();
vectorLayer_tiles.getSource().setTileLoadFunction(function (tile, url) {
  console.log(url);
  return defaultTileLoadFunction(tile, url);
});
```
